### PR TITLE
fix(SelfAccessorFixer): do not touch references inside lambda and/or arrow function

### PR DESCRIPTION
**Pull request by SpacePossum**
closes https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7163

might be best to review per commit, can extract into multiple PRs but not sure worth the trouble, let me know

### CODE REVIEW

File: tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php
Comment (Wirone) [inline tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php:1]: Not a part of this change 🙂.

File: tests/Tokenizer/TokensAnalyzerTest.php
Comment (Wirone) [inline tests/Tokenizer/TokensAnalyzerTest.php:1]: What about:

```php
array-map(
    fn (array $item) => $item['callback']($item['value']),
    [/* items */]
);
```

Wouldn't it treat `]` (closing bracket of first array key fetch) as the last token of that arrow function? There are other scenarios like `fn ($a) => [0, 1, $a]` that wouldn't work as expected if I understand the analyzer correctly.

File: src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php:1]: This call was made ages ago, not sure we want to discuss again, I can revert if you want, leaving a mixed usage of the two once more.

[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php:1]: >PR and should be done in separate one

it is in a separate commit, I think it is fair to ask the reviewers to go by commit, it is not that difficult

File: tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php
Comment (SpacePossum) [reply tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php:1]: it is, by definition, I mean it is in the PR so part of the change, not sure how to read comments like this TBH

File: tests/Tokenizer/TokensAnalyzerTest.php
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:1]: the last index should the `)` before the `,` ,  I can add this case to the tests

File: tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php
Comment (SpacePossum) [reply tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php:1]: also, the description of the PR states:

>might be best to review per commit, can extract into multiple PRs but not sure worth the trouble, let me know

I mean...

**Comment by SpacePossum** [comment]
rebased on latest

File: tests/Tokenizer/TokensAnalyzerTest.php
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:1]: add here https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7349/commits/04dc69b3c4d96617282bd1debf9df38e3a406a25

File: src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php
[the remark this answers is not in the document]
Comment (Wirone) [reply src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php:1]: I don't have any problem with reviewing as a whole or by commits, but that's not the point. When merged, all the changes are squashed into one commit, so reverting only actual feature or only naming change is impossible 🙂. That's why I like respecting SRP on PR level and not mixing things that are not related. I understand the need of changing the naming and I saw there was already such commit in the Git history (made by you) and I'm totally fine with it, I just think this should go as separate PR.

File: tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php
Comment (Wirone) [reply tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php:1]: PR is related to `SelfAccessorFixer` and lambda function, changes in this file are not related with this in any way 🙂. Again, [SRP on PR level](#r1354953955).

File: src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/FunctionNotation/CombineNestedDirnameFixer.php:1]: check, dropped the commits

File: tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php
Comment (SpacePossum) [reply tests/Fixer/FunctionNotation/SingleLineThrowFixerTest.php:1]: cool, dropped

**Comment by SpacePossum** [comment]
rebased on latest, no other changes

**Comment by SpacePossum** [comment]
rebased on latest, no other chatges,
this is a bug fix where a "non-risky" fixer actual _breaks_ peoples code, so maybe someone can take a look? thanks

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 1]
```diff
@@ -2802,6 +2802,151 @@ public function testGetClassyModifiersForNonClassyThrowsAnException(): void
...
+            [
+                2 => 10,
+                13 => 21,
+                24 => 35,
+                42 => 51,
+            ],
+            '<?php
+                fn(&$x) => $x;
+                fn&($x) => $x;
+                fn($x, ...$rest) => $rest;
```
Comment (Wirone) [inline tests/Tokenizer/TokensAnalyzerTest.php:2853]: This is not a reference, it should probably go to `simple cases` test case. Or maybe you meant `fn($x, &...$rest) => $rest;`?

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 2]
```diff
@@ -2802,6 +2802,151 @@ public function testGetClassyModifiersForNonClassyThrowsAnException(): void
...
+                33 => 41,
+            ],
+            '<?php fn($x) => $a = new class($x) { public function foo() { return fn(&$x) => $x; } };',
+        ];
+
+        yield 'array destructuring' => [
+            [
+                4 => 13,
+            ],
+            '<?php return [fn(array $x) => $x1] = $x;',
```
Comment (Wirone) [inline tests/Tokenizer/TokensAnalyzerTest.php:2917]: Is it actually a valid PHP code? It throws `Fatal error: Assignments can only happen to writable values`, but maybe I'm doing something wrong?

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 3]
```diff
@@ -2802,6 +2802,151 @@ public function testGetClassyModifiersForNonClassyThrowsAnException(): void
...
+        ];
+
+        yield [
```
Comment (Wirone) [inline tests/Tokenizer/TokensAnalyzerTest.php:2920]:
```suggestion
        yield 'array_map() callback with different token blocks' => [
```

If I'm guessing the intent correctly 😉.

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 4]
```diff
@@ -2802,6 +2802,151 @@ public function testGetClassyModifiersForNonClassyThrowsAnException(): void
...
+            ],
+            '<?php
+                $a = array_map(
+                    fn (array $item) => $item[\'callback\']($item[\'value\']),
+                    [/* items */]
+                );
+            ',
+        ];
+
+        yield [
```
Comment (Wirone) [inline tests/Tokenizer/TokensAnalyzerTest.php:2932]:
```suggestion
        yield 'arrow function returning array' => [
```
I believe this is again about testing `[]` block? At first I thought it's about assigning to variable, but it wouldn't make sense in terms of finding last token _inside_ arrow function, so maybe that assignment is not needed at all?

**Review by Wirone — CHANGES_REQUESTED** [review]
Some minor comments related to tests, but probably only one can be considered as a blocker (test case with array destructuring). Anyway, good job, let's finish it 🙂.

File: tests/Tokenizer/TokensAnalyzerTest.php (hunk 2 above)
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:2917]: What is valid syntax and what makes sense @ runtime are two different worlds , especially in PHP. The tests passing actually proves it is valid syntax (for now, might be nice if PHP patches it out one day).

File: tests/Tokenizer/TokensAnalyzerTest.php [hunk 5]
```diff
@@ -2802,6 +2802,151 @@ public function testGetClassyModifiersForNonClassyThrowsAnException(): void
...
+        $tokens = Tokens::fromCode($source);
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
+
+        $indices = [];
+
+        foreach ($expectations as $index => $expectedEndIndex) {
+            $indices[$index] = $tokensAnalyzer->getLastTokenIndexOfArrowFunction($index);
+        }
+
+        self::assertSame($expectations, $indices);
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:2821]: I think this works just fine, I know people like to bump the asserting count here (like changing valid `\assert(x)` into `self::assertX(X)`), which is fine, but the assert count is just a number, it does make the test itself any better or worse.

File: tests/Tokenizer/TokensAnalyzerTest.php (hunk 1 above)
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:2854]: Thanks, updated the description and added `fn($x, &...$rest) => 1` as case

File: tests/Tokenizer/TokensAnalyzerTest.php (hunk 3 above)
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php]: thanks, done :)

File: tests/Tokenizer/TokensAnalyzerTest.php (hunk 4 above)
Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php]: thanks, added the description

File: tests/Tokenizer/TokensAnalyzerTest.php (hunk 2 above)
Comment (Wirone) [reply tests/Tokenizer/TokensAnalyzerTest.php:2919]: I'm not in favor of covering cases that are _not_ runnable scenarios, IMHO it's bloating contract between us and users. But it's not something I would fight for 😉.

Comment (SpacePossum) [reply tests/Tokenizer/TokensAnalyzerTest.php:2919]: unittest are not contracts IMHO, but lets do that discussion again :+1:

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you @SpacePossum 🍻

**Pull request merged**